### PR TITLE
Attribute parsing

### PR DESCRIPTION
**Pull request by DenverM80**
Move StructMember centric functions to a new more appropriate StructMemberHelper class

Add boolean property isAttribute to StructMember

Implement generateResponseAttributesParser, Similar to generateResponseParser but handles attribute properties of an XML node rather than name / value of a child node.

Also simplify some logic related to generation of top-level vs embedded properties

### CODE REVIEW

File: ds3-autogen-c/src/main/java/com/spectralogic/ds3autogen/c/models/StructMember.java [hunk 1]
```diff
@@ -18,12 +18,15 @@
 public class StructMember {
     private final C_Type type;
     private final String name;
+    private final boolean isAttribute;
 
     public StructMember(
             final C_Type type,
-            final String name) {
+            final String name,
+            final boolean isAttribute) {
```
Comment (rpmoore) [inline ds3-autogen-c/src/main/java/com/spectralogic/ds3autogen/c/models/StructMember.java]: I would keep a 2 attribute constructor that defaults `isAttribute` to false to clean up some of the test code bellow.

Comment (DenverM80) [inline ds3-autogen-c/src/main/java/com/spectralogic/ds3autogen/c/models/StructMember.java]: and then have a public setter along with the getter?

**Comment by DenverM80** [comment]
updated,  parsing related issues have been fixed.

File: ds3-autogen-c/src/main/java/com/spectralogic/ds3autogen/c/helpers/StructMemberHelper.java [hunk 2]
```diff
@@ -0,0 +1,203 @@
+package com.spectralogic.ds3autogen.c.helpers;
```
Comment (RachelTucker) [inline ds3-autogen-c/src/main/java/com/spectralogic/ds3autogen/c/helpers/StructMemberHelper.java]: missing copyright

File: ds3-autogen-c/src/main/java/com/spectralogic/ds3autogen/c/helpers/StructMemberHelper.java [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+import com.google.common.collect.ImmutableList;
+import com.spectralogic.ds3autogen.c.models.StructMember;
+import com.spectralogic.ds3autogen.utils.Helper;
+import com.spectralogic.ds3autogen.utils.collections.GuavaCollectors;
+
+import java.text.ParseException;
+
+import static com.spectralogic.ds3autogen.utils.Helper.indent;
+
+public class StructMemberHelper {
```
Comment (RachelTucker) [inline ds3-autogen-c/src/main/java/com/spectralogic/ds3autogen/c/helpers/StructMemberHelper.java]: class should be final

File: ds3-autogen-c/src/main/java/com/spectralogic/ds3autogen/c/helpers/StructHelper.java [hunk 4]
```diff
@@ -45,6 +45,23 @@ public static String getNameUnderscores(final String name) {
...
+     * @return camel-cased structMemberName
+     */
+    public static String convertStructMemberNameToCamelCase(final String structName, final String structMemberName) {
+        if (structMemberName.equals("id")) {
+            if (structName.equals("ds3_user_response")) {
+                return "Id";
+            }
+            return "ID";
+        }
+        //return Helper.underscoreToCamel(structMemberName);
```
Comment (shabtaisharon) [inline ds3-autogen-c/src/main/java/com/spectralogic/ds3autogen/c/helpers/StructHelper.java]: can be removed?

File: ds3-autogen-c/src/main/java/com/spectralogic/ds3autogen/c/helpers/StructMemberHelper.java [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+                .filter(sm -> !sm.hasWrapper())
+                .collect(GuavaCollectors.immutableList());
+    }
+
+    public static String generateStructMemberParserLine(final StructMember structMember, final String parserFunction) throws ParseException {
+        return indent(3) + "response->" + Helper.camelToUnderscore(structMember.getName()) + " = " + parserFunction + "\n";
+    }
+
+    public static String generateStructMemberDs3StrArrayBlock(final StructMember structMember) {
+        return indent(3) + "xmlNodePtr loop_node;\n"
```
Comment (shabtaisharon) [inline ds3-autogen-c/src/main/java/com/spectralogic/ds3autogen/c/helpers/StructMemberHelper.java]: do you think you can const those numbers and give them a name, like why 3 or 4?

Comment (rpmoore) [inline ds3-autogen-c/src/main/java/com/spectralogic/ds3autogen/c/helpers/StructMemberHelper.java]: it seems like this whole string could be made const.

File: ds3-autogen-c/src/main/java/com/spectralogic/ds3autogen/c/helpers/StructHelper.java (hunk 4 above)
Comment (DenverM80) [inline ds3-autogen-c/src/main/java/com/spectralogic/ds3autogen/c/helpers/StructHelper.java]: I removed this whole funtion in the next pull request when I add the nameToMarshall property to the StructMember

File: ds3-autogen-c/src/main/java/com/spectralogic/ds3autogen/c/helpers/StructMemberHelper.java (hunk 3 above)
Comment (DenverM80) [inline ds3-autogen-c/src/main/java/com/spectralogic/ds3autogen/c/helpers/StructMemberHelper.java]: fixed in next pull request

File: ds3-autogen-c/src/main/java/com/spectralogic/ds3autogen/c/helpers/StructMemberHelper.java (hunk 2 above)
Comment (DenverM80) [inline ds3-autogen-c/src/main/java/com/spectralogic/ds3autogen/c/helpers/StructMemberHelper.java]: fixed in next pull request

File: ds3-autogen-c/src/main/java/com/spectralogic/ds3autogen/c/helpers/StructMemberHelper.java (hunk 5 above)
Comment (DenverM80) [inline ds3-autogen-c/src/main/java/com/spectralogic/ds3autogen/c/helpers/StructMemberHelper.java]: That's how the code lines up with the template that calls it (/ds3_autogen/ds3-autogen-c/src/main/resources/templates/source-templates/ResponseParser.ftl) ... I want to move these to small templates at some point very soon but have been working through higher priority problems

**Comment by DenverM80** [comment]
updated.

**Comment by DenverM80** [comment]
Bump

**Pull request merged**